serialization/python_pickle.ksy

```diff
@@ -0,0 +1,588 @@
+          are only available in later versions of the Pickle protocol.
+      - id: arg
+        type:
+          switch-on: code
+          cases:
+            'opcode::int': decimalnl_short
+            'opcode::binint': s4
+            'opcode::binint1': u1
+            'opcode::binint2': u2
+            'opcode::long': decimalnl_long
+            'opcode::long1': long1
+            'opcode::long4': long4
+            'opcode::string': stringnl
+            'opcode::binstring': string4
+            'opcode::short_binstring': string1
+            'opcode::binbytes': bytes4
+            'opcode::short_binbytes': bytes1
+            'opcode::binbytes8': bytes8
+            'opcode::none': no_arg
+            'opcode::newtrue': no_arg
+            'opcode::newfalse': no_arg
+            'opcode::unicode': unicodestringnl
+            'opcode::short_binunicode': unicodestring1
+            'opcode::binunicode': unicodestring4
+            'opcode::binunicode8': unicodestring8
+            'opcode::float': floatnl
+            'opcode::binfloat': f8
+            'opcode::empty_list': no_arg
+            'opcode::append': no_arg
+            'opcode::appends': no_arg
+            'opcode::list': no_arg
+            'opcode::empty_tuple': no_arg
+            'opcode::tuple': no_arg
+            'opcode::tuple1': no_arg
+            'opcode::tuple2': no_arg
+            'opcode::tuple3': no_arg
+            'opcode::empty_dict': no_arg
+            'opcode::dict': no_arg
+            'opcode::setitem': no_arg
+            'opcode::setitems': no_arg
+            'opcode::empty_set': no_arg
+            'opcode::additems': no_arg
+            'opcode::frozenset': no_arg
+            'opcode::pop': no_arg
+            'opcode::dup': no_arg
+            'opcode::mark': no_arg
+            'opcode::pop_mark': no_arg
+            'opcode::get': decimalnl_short
+            'opcode::binget': u1
+            'opcode::long_binget': u4
+            'opcode::put': decimalnl_short
+            'opcode::binput': u1
+            'opcode::long_binput': u4
+            'opcode::memoize': no_arg
+            'opcode::ext1': u1
+            'opcode::ext2': u2
+            'opcode::ext4': u4
+            'opcode::global':  stringnl_noescape_pair
+            'opcode::stack_global': no_arg
+            'opcode::reduce': no_arg
+            'opcode::build': no_arg
+            'opcode::inst': stringnl_noescape_pair
+            'opcode::obj': no_arg
+            'opcode::newobj': no_arg
+            'opcode::newobj_ex': no_arg
+            'opcode::proto': u1
+            'opcode::stop': no_arg
+            'opcode::frame': u8
+            'opcode::persid': stringnl_noescape
+            'opcode::binpersid': no_arg
+        doc: |
+          Optional argument for the operation. Data type and length
+          are determined by the value of the opcode.
+
+  decimalnl_short:
+    seq:
+      - id: val
+        type: str
+        encoding: ASCII
```

Comment:
6 `ASCII` vs 3 `utf8`. May make sense to move `ascii` into `meta`.

Reply:
> 6 `ASCII` vs 3 `utf8`. May make sense to move `ascii` into `meta`.

Hm, I think in this case it's better to put the encodings on the individual fields. I would use a global `encoding` only if there is one obvious "main" encoding with very few or no exceptions. A ratio of 2:1 is not clear enough IMHO, and since there aren't *that* many string fields in the spec, it wouldn't make a big difference in terms of compactness.

> BTW, the convention is to use lowercase in encodings.

Is it? I searched through the ksf repo and found both uppercase and lowercase used often, with uppercase being a bit more common than lowercase. All the examples in the [user guide](http://doc.kaitai.io/user_guide.html) use uppercase encoding names too. The [style guide](http://doc.kaitai.io/ksy_style_guide.html) doesn't say anything about how encoding names should be written, maybe we should change that?

Reply:
>The style guide doesn't say anything about how encoding names should be written, maybe we should change that?

Yes, we should. The discussion about encoding casing is https://github.com/kaitai-io/kaitai_struct/issues/391

Reply:
I'll use lowercase throughout.